### PR TITLE
tests: update for tenant generations

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Some existing tests are written in a way that's incompatible with tenant generations.

## Summary of changes

Update all the tests that need updating: this is things like calling through the NeonPageserver.tenant_attach helper to get a generation number, instead of calling directly into the pageserver API.  There are various more subtle cases.

### CODE REVIEW

**Comment by jcsp** [comment]
Interesting failure of test_broker:
- https://neon-github-public-dev.s3.amazonaws.com/reports/pr-5449/7102616600/index.html#suites/82004ab4e3720b47bf78f312dabe7c55/dcc92f8400d731e4
- the test is timing out waiting for remote_consistent_lsn to advance.  We can see the pageserver doing a successful validation call to attachment service.  The pageserver should then be sending the updated remote_consistent_lsn in response to a heartbeat from the safekeeper.

File: libs/remote_storage/src/local_fs.rs [hunk 1]
```diff
@@ -99,27 +99,34 @@ impl LocalFs {
...
             }
-            Err(e) if e.kind() == ErrorKind::NotFound => {
-                // It's not a file that exists: strip the prefix back to the parent directory
-                initial_dir.pop();
-            }
-            Err(e) => {
-                // Unexpected I/O error
-                anyhow::bail!(e)
+
+            if initial_dir.parent().is_none() {
```
Comment (arpad-m) [inline libs/remote_storage/src/local_fs.rs:126]: given this `if`, I wonder if we ever hit the case where we exit the while loop due to `initial_dir == Utf8PathBuf::default()`. Then, we might just have a `loop` here, no?

File: test_runner/fixtures/pageserver/http.py [hunk 2]
```diff
@@ -210,16 +210,25 @@ def tenant_list(self) -> List[Dict[Any, Any]]:
         return res_json
 
     def tenant_create(
-        self, new_tenant_id: TenantId, conf: Optional[Dict[str, Any]] = None
+        self,
+        new_tenant_id: TenantId,
+        conf: None | Dict[str, Any] = None,
```
Comment (arpad-m) [inline test_runner/fixtures/pageserver/http.py:215]: Why did you change this to `None |` instead of `Optional[Dict[..]]`? I don't know, I am not familiar with python enough to know the answer.

File: test_runner/fixtures/remote_storage.py [hunk 3]
```diff
@@ -88,13 +87,49 @@ def tenant_path(self, tenant_id: TenantId) -> Path:
...
+    ):
+        if generation is None:
+            generation = self.timeline_latest_generation(tenant_id, timeline_id)
+
+        if generation is None:
+            filename = local_name
+        else:
+            filename = f"{local_name}-{generation:08x}"
+
+        # Assume this is a simple test that uploads in just one generation
```
Comment (arpad-m) [inline test_runner/fixtures/remote_storage.py:131]: is this comment still accurate? the code is generation aware now, no?

File: test_runner/regress/test_tenant_relocation.py [hunk 4]
```diff
@@ -40,58 +34,6 @@ def assert_abs_margin_ratio(a: float, b: float, margin_ratio: float):
     assert abs(a - b) / a < margin_ratio, abs(a - b) / a
 
 
-@contextmanager
-def new_pageserver_service(
```
Comment (arpad-m) [inline test_runner/regress/test_tenant_relocation.py:44]: just to make sure: this is not a test but it is a helper function that is now redundant? python syntax is a little implicit for me to figure it out for sure just from looking at the function definition.

File: test_runner/regress/test_wal_acceptor.py [hunk 5]
```diff
@@ -298,17 +298,21 @@ def test_broker(neon_env_builder: NeonEnvBuilder):
...
-    # deactivated once rcl reaches pageserver one, and thus we generally wait
-    # till pageserver reconnects to all safekeepers one by one here. Timeline
-    # status on safekeeper should take into account peers state as well.
+    # This timeout is long: safekeepers learn about remote_consistent_lsn updates when a pageserver
+    # connects, receives a PrimaryKeepAlive, and sends a PageserverFeedback.  So the timeout has to encompass:
+    # - pageserver deletion_queue to validate + publish the remote_consistent_lsn
+    # - pageserver to reconnect to all safekeepers one by one, with multi-second delays between
+    #
+    # TODO: timeline status on safekeeper should take into account peers state as well.
+    rcl_propagate_secs = 60
```
Comment (arpad-m) [inline test_runner/regress/test_wal_acceptor.py:307]: you doubled this in this PR. Do generations make this test slower or is it just to make the test less flaky?

**Review by arpad-m — CHANGES_REQUESTED** [review]

File: test_runner/regress/test_wal_acceptor.py (hunk 5 above)
Comment (jcsp) [reply test_runner/regress/test_wal_acceptor.py:307]: Yes, the runtime is longer now: there is a delay between writing to remote storage and validating our update to remote_consistent_lsn.  The test was previously close to its 30 second timeout and this could occasionally push it over the edge.

File: libs/remote_storage/src/local_fs.rs (hunk 1 above)
Comment (jcsp) [reply libs/remote_storage/src/local_fs.rs:126]: Yes, this can be simplified (we also have to pull the parent.is_none check to the top of the loop to make this work) 9c0353410

File: test_runner/fixtures/pageserver/http.py (hunk 2 above)
Comment (jcsp) [reply test_runner/fixtures/pageserver/http.py:215]: Hmm, I imagine that I was working around some issue with the linter when I wrote this, but when I try reverting this to Optional the linter is also happy.  

Switched back to optional in f0eda5429

File: test_runner/fixtures/remote_storage.py (hunk 3 above)
Comment (jcsp) [reply test_runner/fixtures/remote_storage.py:131]: Yeah, the comment is outdated.  I think we can probably also entirely remove the handling for the case where we can't figure out the generation, since all tests will be uploading layers with generation suffixes: it doesn't make sense to try and fall back to generation-less paths.  b0bf75f6e

File: test_runner/regress/test_tenant_relocation.py (hunk 4 above)
Comment (jcsp) [reply test_runner/regress/test_tenant_relocation.py:44]: Right.  This test used to have some special code for constructing additional pageserver instances, but since NeonEnv got the ability to run multiple pageservers, it is no longer needed.  I cleaned it up in this PR so that the test would have accesss to proper NeonPageserver objects, as their generation-aware helpers are now needed.

**Review by arpad-m — APPROVED** [review]
